sql/database.go

```diff
@@ -444,6 +683,177 @@ func (db *sqliteDatabase) Close() error {
 	return nil
 }
 
+// Intercept adds an interceptor function to the database. The interceptor functions
+// are invoked upon each query. The query will fail if the interceptor returns an error.
+// The interceptor can later be removed using RemoveInterceptor with the same key.
+func (db *sqliteDatabase) Intercept(key string, fn Interceptor) {
+	db.interceptMtx.Lock()
+	defer db.interceptMtx.Unlock()
+	if db.interceptors == nil {
+		db.interceptors = make(map[string]Interceptor)
+	}
+	db.interceptors[key] = fn
+}
+
+// PopIntercept removes the interceptor function with specified key from the database.
+// If there's no such interceptor, the function does nothing.
+func (db *sqliteDatabase) RemoveInterceptor(key string) {
+	db.interceptMtx.Lock()
+	defer db.interceptMtx.Unlock()
+	delete(db.interceptors, key)
+}
+
+// vacuumInto runs VACUUM INTO on the database and saves the vacuumed
+// database at toPath.
+func (db *sqliteDatabase) vacuumInto(toPath string) error {
+	if _, err := db.Exec("VACUUM INTO ?1", func(stmt *Statement) {
+		stmt.BindText(1, toPath)
+	}, nil); err != nil {
+		return fmt.Errorf("vacuum into %s: %w", toPath, err)
+	}
+	return nil
+}
+
+// copyMigrateDB performs a copy-based migration of the database.
+// The source database is always closed by this function.
+// Upon success, the migrated database is opened.
+func (db *sqliteDatabase) copyMigrateDB(config *conf) (finalDB *sqliteDatabase, err error) {
+	defer db.Close()
+
+	dbPath, migratedPath, err := dbMigrationPaths(config.uri)
+	if err != nil {
+		return nil, err
+	}
+	if migratedPath == "" {
+		return nil, fmt.Errorf("cannot migrate database, only file DBs are supported: %s", config.uri)
+	}
+
+	// Instead of just copying the source database to the temporary migration DB, use VACUUM INTO.
+	// This is somewhat slower but achieves two goals:
+	// 1. The lock is held on the source database while it's being copied
+	// 2. If the source database has a lot of free pages for whatever reason, those
+	// are not copied, saving disk space
+	config.logger.Info("making a temporary copy of the database",
+		zap.String("path", dbPath),
+		zap.String("target", migratedPath))
+	if err := db.vacuumInto(migratedPath); err != nil {
+		return nil, errors.Join(err, deleteDB(migratedPath))
```

Comment:
I don't think we need to use `errors.Join` here. It might even lead to confusing errors where it's unclear which error was caused by VACUUM failing and which one by not being able to delete the DB:

```suggestion
		deleteDB(migratedPath)
		return nil, err
```

Reply:
`deleteDB` is quite a bit different from other situations with cleanup errors.
Imagine the following situation: a transient disk access failure happens on Linux, causing the filesystem to become read-only (say a loose SATA cable on an old SSD), and the migration fails. Afterwards, the temporary or old DB cannot be deleted as the fs is still read-only, and this causes a 150GiB+ file to remain in place, occupying some significant disk space. The user may or may not have intent to run `go-spacemesh` again after recovering from the problem, and if she/he doesn't, the file will not be deleted unless the user does it by hand.
I'd say that this huge file is something we better tell the user about.

Also, if `deleteDB` fails, it returns an error of the form
```remove /path/to/database/file: <actual error such as read-only filesystem>```
This is rather distinct from the other part of the error message and is rather easy to understand IMO.